src/k2/cmd/ycsb/data.h

```diff
@@ -0,0 +1,155 @@
+static const String ycsbCollectionName = "YCSB";
+
+// helper function to convert to base (default 93 (ascii from 33 to 126)) representation
+String itostring(int n, int base = 93)
+{
+    String buf(String::initialized_later{}, 1000);
+    std::div_t dv{}; dv.quot = n;
+    uint32_t i = 0;
+
+    do {
+        dv = std::div(dv.quot, base);
+        buf[i] = (char)(std::abs(dv.rem)+33);  // avoid ascii value 126 because it is used for padding
+        i++;
+    } while(dv.quot);
+
+    return buf.substr(0,i);
+}
+
+class YCSBData{
+
+public:
+
+    static dto::Schema generate_schema(const uint32_t num_fields){
+
+        std::vector<dto::SchemaField> field_names;
+
+        for(uint32_t field_no = 0; field_no < num_fields; ++field_no){
+            String fieldname = "field" + std::to_string(field_no);
+            dto::SchemaField f {dto::FieldType::STRING, fieldname , false, false};
+            field_names.push_back(f);
+        }
+
+        dto::Schema ycsb_data_schema {
+            .name = "ycsb_data",
+            .version = 1,
+            .fields = field_names,
+            .partitionKeyFields = std::vector<uint32_t> { 0 },
+            .rangeKeyFields = std::vector<uint32_t> {}
+        };
+
+        return ycsb_data_schema;
+    }
+
+    static String idToKey(uint64_t keyid, uint32_t len_field){
+        String key = itostring(keyid);
+
+        if(key.size()>len_field)
+            return "";
+
+        String padding( (len_field - key.size()),char(126));
+
+        return padding + key;
+    }
+
+    YCSBData(uint64_t keyid, bool generateNew = true) : ID(keyid) {
+
+        fields.push_back(YCSBData::idToKey(keyid,_len_field()));
+
+        if(fields[0]=="" || !generateNew) // key = "" means id cannot be translated to string of required length
+            return;
+
+        RandomContext random(0);
+
+        for(uint32_t field_no = 1; field_no < _num_fields(); ++field_no){
+             fields.push_back(random.RandomString(_len_field()));
+        }
+    }
+
+    uint64_t ID;
+    std::vector<String> fields;
+    static inline dto::Schema ycsb_schema;
+    static inline thread_local std::shared_ptr<dto::Schema> schema;
+    static inline String collectionName = ycsbCollectionName;
+
+private:
+    k2::ConfigVar<uint32_t> _len_field{"field_length"};
+    k2::ConfigVar<uint32_t> _num_fields{"num_fields"};
+};
+
+seastar::future<WriteResult> writeRow(YCSBData& row, K2TxnHandle& txn, bool erase = false)
+{
+    dto::SKVRecord skv_record(YCSBData::collectionName, YCSBData::schema); // create SKV record
+
+    for(auto field : row.fields){
+        skv_record.serializeNext<String>(field); // add fields to SKV record
+    }
+
+    return txn.write<dto::SKVRecord>(skv_record, erase).then([] (WriteResult&& result) {
+        if (!result.status.is2xxOK()) {
+            K2LOG_D(log::ycsb, "writeRow failed: {}", result.status);
+            return seastar::make_exception_future<WriteResult>(std::runtime_error("writeRow failed!"));
```

Comment:
I'm not sure why you're converting all statuses to exception. It seems to me this is fine for load, but for regular txn writes, you'd want to handle say conflicts better?

Reply:
This is how I do it in TPC-C. The idea is that if any operation fails in a transaction, it is propagated up to the transaction runner via exception which simply aborts the transaction and marks it as failed for metrics.